### PR TITLE
improve Vec example soundness in mem::transmute docs

### PR DESCRIPTION
**Pull request by llogiq**
The previous version of the `Vec` example had a case of questionable soundness, because at one point `v_orig` was aliased.

r? @RalfJung

### CODE REVIEW

File: src/libcore/intrinsics.rs [hunk 1]
```diff
@@ -869,11 +869,12 @@ extern "rust-intrinsic" {
     /// // the original inner type (`&i32`) to the converted inner type
     /// // (`Option<&i32>`), so read the nomicon pages linked above.
     /// let v_from_raw = unsafe {
+    ///     // ensure the original vector is not dropped
```
Comment (Centril) [inline src/libcore/intrinsics.rs]:
```suggestion
    ///     // Ensure the original vector is not dropped.
```

Comment (llogiq) [reply src/libcore/intrinsics.rs]: Thanks @Centril!

File: src/libcore/intrinsics.rs [hunk 2]
```diff
@@ -869,11 +869,12 @@ extern "rust-intrinsic" {
     /// // the original inner type (`&i32`) to the converted inner type
     /// // (`Option<&i32>`), so read the nomicon pages linked above.
     /// let v_from_raw = unsafe {
+    ///     // Ensure the original vector is not dropped
```
Comment (Centril) [inline src/libcore/intrinsics.rs]:
```suggestion
    ///     // Ensure the original vector is not dropped.
```
You missed the punctuation. ;)

Comment (llogiq) [reply src/libcore/intrinsics.rs]: Sorry. I added it. Thanks again.

**Comment by RalfJung** [comment]
> The previous version of the Vec example had a case of questionable soundness, because at one point v_orig was aliased.

Nice! Yes `ManuallyDrop` is certainly better.

However, the code is still somewhat confusing, with its three examples in one doctest and the `clone` that do not have anything to do with the actual operation.
Maybe it would be worth refactoring it to have one block per "way" of doing this transmute, and have each block look like
```rust
{
    let v = v_orig.clone();
    // Here comes the "way" do do stuff with `v`.
}
```
That would make it more clear that the clone is part of the "environment" here, not part of what is really going on.

However, we can also land this PR without that change as it is a definite improvement.

File: src/libcore/intrinsics.rs [hunk 3]
```diff
@@ -869,11 +872,12 @@ extern "rust-intrinsic" {
     /// // the original inner type (`&i32`) to the converted inner type
     /// // (`Option<&i32>`), so read the nomicon pages linked above.
     /// let v_from_raw = unsafe {
+    ///     // Ensure the original vector is not dropped.
+    ///     let v_orig: &mut Vec<_> = &mut *std::mem::ManuallyDrop::new(v_orig);
```
Comment (RalfJung) [inline src/libcore/intrinsics.rs]: This is still not consistently using the same pattern for all three cases. For uniformity, the last example IMO should not exploit the fact that it is indeed the last example.

So, I think there should be a third `v_clone`.

Comment (llogiq) [reply src/libcore/intrinsics.rs]: Makes sense. Wait for it...

File: src/libcore/intrinsics.rs [hunk 4]
```diff
@@ -847,19 +847,24 @@ extern "rust-intrinsic" {
     /// let store = [0, 1, 2, 3];
     /// let mut v_orig = store.iter().collect::<Vec<&i32>>();
     ///
+    /// // clone the references as we will reuse them later
```
Comment (RalfJung) [inline src/libcore/intrinsics.rs]: Why not "clone the vector"?

File: src/libcore/intrinsics.rs [hunk 5]
```diff
@@ -869,11 +874,12 @@ extern "rust-intrinsic" {
     /// // the original inner type (`&i32`) to the converted inner type
     /// // (`Option<&i32>`), so read the nomicon pages linked above.
     /// let v_from_raw = unsafe {
-    ///     Vec::from_raw_parts(v_orig.as_mut_ptr() as *mut Option<&i32>,
-    ///                         v_orig.len(),
-    ///                         v_orig.capacity())
+    ///     // Ensure the original vector is not dropped.
+    ///     let v_clone: &mut Vec<_> = &mut *std::mem::ManuallyDrop::new(v_clone);
```
Comment (RalfJung) [inline src/libcore/intrinsics.rs]:
```suggestion
    ///     let v_clone = std::mem::ManuallyDrop::new(v_clone);
```
This should work due to the `Deref`, right? And I think it is more readable.

**Comment by llogiq** [comment]
@RalfJung thanks for the review!

**Comment by RalfJung** [comment]
LGTM, thanks!

@bors r+ rollup

**Comment by RalfJung** [comment]
@bors r- delegate+
Please fix the CI failure, wait for CI in the PR to be green, then do `bors r=RalfJung`

**Pull request merged**